orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/handler/ContinueParentStageHandler.kt

```diff
@@ -46,14 +43,23 @@ class ContinueParentStageHandler(
 
   override fun handle(message: ContinueParentStage) {
     message.withStage { stage ->
-      if (stage.allBeforeStagesComplete()) {
-        when {
-          stage.hasTasks() -> stage.runFirstTask()
-          else -> queue.push(CompleteStage(stage))
+      if (message.phase == STAGE_BEFORE) {
+        if (stage.allBeforeStagesComplete()) {
+          when {
+            stage.hasTasks() -> stage.runFirstTask()
+            else             -> queue.push(CompleteStage(stage))
+          }
+        } else if (!stage.anyBeforeStagesFailed()) {
+          log.warn("Re-queuing $message as other ${message.phase} stages are still running")
+          queue.push(message, retryDelay)
+        }
+      } else {
+        if (stage.allAfterStagesComplete()) {
+          queue.push(CompleteStage(stage))
+        } else if (!stage.anyAfterStagesFailed()) {
+          log.warn("Re-queuing $message as other ${message.phase} stages are still running")
+          queue.push(message, retryDelay)
```

Comment:
This adds the capability for after stages to be parallel (with multiple end points) which we didn't support before. I didn't end up using it, but it will probably avoid someone getting confused about why it doesn't work later down the line.